contracts/modules/Experimental/Checkpoints/WeightedVoteCheckpoint.sol

```diff
@@ -0,0 +1,216 @@
+     * @param _noOfProposals Number of proposals
+     */
+    function createBallot(uint256 _duration, uint256 _noOfProposals) external withPerm(ADMIN) {
+        require(_duration > 0, "Incorrect ballot duration.");
+        uint256 checkpointId = ISecurityToken(securityToken).createCheckpoint();
+        uint256 endTime = now.add(_duration);
+        _createCustomBallot(now, endTime, checkpointId, _noOfProposals);
+    }
+
+    function _createCustomBallot(uint256 _startTime, uint256 _endTime, uint256 _checkpointId, uint256 _noOfProposals) internal {
+        require(_noOfProposals > 1, "Incorrect proposals no");
+        require(_endTime > _startTime, "Times are not valid");
+        uint256 ballotId = ballots.length;
+        uint256 supplyAtCheckpoint = ISecurityToken(securityToken).totalSupplyAt(_checkpointId);
+        ballots.push(
+            Ballot(
+                _checkpointId, supplyAtCheckpoint, uint64(_startTime), uint64(_endTime), uint64(0), uint56(_noOfProposals), uint8(1)
+            )
+        );
+        emit BallotCreated(_startTime, _endTime, ballotId, _checkpointId, _noOfProposals);
+    }
+
+    /**
+     * @notice Allows the token issuer to create a ballot with custom settings
+     * @param _startTime Start time of the voting period in Unix Epoch time
+     * @param _endTime End time of the voting period in Unix Epoch time
+     * @param _checkpointId Index of the checkpoint to use for token balances
+     * @param _noOfProposals Number of proposals
+     */
+    function createCustomBallot(uint256 _startTime, uint256 _endTime, uint256 _checkpointId, uint256 _noOfProposals) external withPerm(ADMIN) {
+        require(_checkpointId <= ISecurityToken(securityToken).currentCheckpointId(), "Invalid checkpoint Id");
+        _createCustomBallot(_startTime, _endTime, _checkpointId, _noOfProposals);
+    }
+
+    /**
+     * @notice Allows a token holder to cast their vote on a specific ballot
+     * @param _ballotId The index of the target ballot
+     * @param _proposalId Id of the proposal which investor want to vote for proposal
+     */
+    function castVote(uint256 _ballotId, uint256 _proposalId) external {
+        require(_ballotId < ballots.length, "Incorrect ballot Id");
+        Ballot storage ballot = ballots[_ballotId];
+
+        uint256 weight = ISecurityToken(securityToken).balanceOfAt(msg.sender, ballot.checkpointId);
+        require(weight > 0, "weight should be > 0");
+        require(ballot.totalProposals >= _proposalId && _proposalId > 0, "Incorrect proposals Id");
+        require(now >= ballot.startTime && now <= ballot.endTime, "Voting period is not active");
+        require(ballot.voteByAddress[msg.sender] == 0, "Token holder has already voted");
+        require(ballot.isActive == uint8(1), "Ballot is not active");
+
+        ballot.voteByAddress[msg.sender] = _proposalId;
+        ballot.totalNumVotes = uint64(uint256(ballot.totalNumVotes).add(uint256(1)));
```

Comment:
Minor optimization: we don't need to use safemath.add here because it can never overflow.
(Adding 1 to a 64 bit number can never result in overflow of a 256 bit number). We don't even need to check for overflow of 64bits because I doubt any ballot is gonna have 18,446,744,073,709,551,616 votes cast :).